sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -342,17 +342,14 @@ class MicroBatchExecution(
         isCurrentBatchConstructed = true
         availableOffsets = nextOffsets.toStreamProgress(sources)
         /* Initialize committed offsets to a committed batch, which at this
-         * is the second latest batch id in the offset log. */
-        if (latestBatchId != 0) {
-          val secondLatestOffsets = offsetLog.get(latestBatchId - 1).getOrElse {
-            logError(s"The offset log for batch ${latestBatchId - 1} doesn't exist, " +
-              s"which is required to restart the query from the latest batch $latestBatchId " +
-              "from the offset log. Please ensure there are two subsequent offset logs " +
-              "available for the latest batch via manually deleting the offset file(s). " +
-              "Please also ensure the latest batch for commit log is equal or one batch " +
-              "earlier than the latest batch for offset log.")
-            throw new IllegalStateException(s"batch ${latestBatchId - 1} doesn't exist")
-          }
+         * is the second latest batch id in the offset log.
```

Comment:
> Do we have a goal to support smooth transition between normal microbatch execution and async progress tracking for a single query?

yes

The existing behavior will break async progress tracking especially if the user wants to switch between turning it on and off

Reply:
I got back here again as I feel like it's not safe to do the change in normal microbatch execution. 

Say, if the user doesn't use async progress tracking at all. (Using sink which is not listed in async progress tracking support or using stateful operator are all valid.) And suppose there is an issue on storage and there are gaps on offset log somehow.

Previously, the query will fail fast. It will stop processing, but it won't break anything.

Now, we allow query to go through, with silently breaking semantic of exactly once. Idempotent write based on batch ID will be broken as offset range for the batch ID gets changed. Same for state store.

Wouldn't it be more serious problem than supporting switch? If we really want to support switching, can we only support switch for the case when checkpoint interval is disabled, so that we don't make change on normal microbatch execution which could lead to break on fault tolerance semantic?

Reply:
This logic is not guaranteeing exactly once behavior.  This logic here is merely a sanity check to guard against bugs.  Not having this check is not breaking exactly once behavior. 


> Wouldn't it be more serious problem than supporting switch? If we really want to support switching, can we only support switch for the case when checkpoint interval is disabled, so that we don't make change on normal microbatch execution which could lead to break on fault tolerance semantic?

I don't quite follow. How would this work? The framework will have to somehow remember the settings of the previous run.  We would need to add metadata to offsets to determine which offsets were written when async progress tracking is used.  We don't have any of this kind of functionality in spark today and not convinced it is worth while to implement such a thing for this use case.

Reply:
How often do we actually catch issues with this check?

Reply:
This logic can affect the offset range of microbatch. As you've added the test, even without async progress tracking flag on, normal processing trigger can technically roll multiple microbatches back, "with composing these offsets into one". This breaks the assumption of exactly-once semantic, every microbatch should have planned its offset range before execution, and the range must not be changed once planned.

This is why async progress tracking cannot work as it is for Delta sink and stateful operator. We blocked this for async progress tracking, but accidentally exposing this to "normal" processing trigger.

Reply:
How often we encounter this is not important. The fact we are removing guard to not break fault-tolerance semantic is important.